### DISCUSSION TITLE
Generic Material property based on Damage Evolution

### DISCUSSION THREAD

**Discussion by batodon** [idaholab/moose#22195] [Q&A Modules: Solid mechanics] [answered]
Hello,

I’m trying to implement the damage evolution shown below as a material property so that another material can call the damage variable (for example, as a damage index). Can anyone recommend an example MOOSE file in this regard? How can I approach this? Thank you. 

![DamageEv](https://user-images.githubusercontent.com/113298585/192885520-20cee7f6-cefe-4dce-b272-56bba3903ea6.PNG)

**Comment by GiudGiud** [comment]
Hello

You should be able to make a new material property for this. 
If you follow the tutorial we do introduce material properties there with an example

What dimension is this quantity? It's a scalar right?

Guillaume

**Reply by batodon** [reply]
Thanks, Guillaume. The damage variable is a scalar quantity and it evolves with time.

**Reply by GiudGiud** [reply]
ok a regular material property (type Real) should be appropriate then.

**Reply by batodon** [reply]
Hello @GiudGiud,

When developing a material property, do you always have to inherit from the material class? Or can you inherit from a different class and compute your _qp properties? For example, I inherited from a “radialreturncreepstressupdate” class because it will enable me to apply the newton iteration to compute my damage variable. However, when I tried using the computeQpProperties method, I got the following Error shown below. (P/S: I attached the files below). 

![damEvol](https://user-images.githubusercontent.com/113298585/192674117-3a6840c2-cf9a-4370-a440-3e7ad5d74f45.PNG)


[damageEvol.zip](https://github.com/idaholab/moose/files/9660935/damageEvol.zip)

**Reply by GiudGiud** [reply]
Hello

Yes you have to at some level. You can inherit from a child of the material class though. 
It works here, radialreturncreepstressupdate inherits from RadialReturnStressUpdate, which inherits from StressUpdateBase, which inherits from Material. 

From the errors it looks like you should be writing HLDamageEvoleTempl<is_ad> (with is_ad = true or false depending on whether you are using AD). 
I think the problem is that you are inheriting from RadialReturnCreepStressUpdateBaseTempl instead of ADRadialReturnCreepStressUpdateBase, which is the instantiation of that template. 

Or is this on purpose? Do you want to be creating a new template or a new class?

Guillaume

**Reply by batodon** [reply]
Hi @GiudGiud,

Thanks. I’m creating a material class (not a template) with a damaged property defined at the _qp points so that another material object may call this _qp damaged property. I added  **template** <**bool is_ad**> on top the definition of the computeQpProperties() method, and it seems to be working now except for one error (shown below).

The error indicated that ‘scalar’ (which is computed within another method in the class) could not be used in the ‘computeQpProperties()’ method without first declaring it. How do I declare it so I can use it to compute the final state of my damaged property?  (P/S: In case you want to see the files, it is attached to my previous question). 

![damEvol2](https://user-images.githubusercontent.com/113298585/192697181-56264a41-52bd-4eb2-8dbd-7875f0f05c2a.PNG)

**Reply by GiudGiud** [reply]
you need to make scalar (call it _scalar instead) a class attribute. You define it in the header as a part of the class, then each routine may use the value and modify it too

**Reply by batodon** [reply]
Hi @GiudGiud,

It didn’t work. However, it worked when I defined my own method/function (without overriding the computeQpProperties method) and declared scalar. I’m not entirely sure. Is this a good code development? Do you think the damage variable is computed successfully as _qp properties? And do you think the scalar obtained from the previous methods is successfully used in this new method/function? See attached for the files 

[damageEvol.zip](https://github.com/idaholab/moose/files/9668764/damageEvol.zip)

**Reply by GiudGiud** [reply]
If you do not override computeQpProperties you are not computing the property, this is clear. 

So no I do not think it s computed successfully. 

I do not see _scalar as a class attribute in your input. 
Should the scalar be a material property instead? What does it represent?

**Reply by batodon** [reply]
Hi @GiudGiud,

So basically, I’m setting up the damage evolution as a residual equation to be computed by newton iteration using the “computeResidualInternal” and “computeDerivative” methods (similar to what is done in the IsotropicPlasticityStressUpdate or PowerLawCreepStressUpdate classes). The scalar is the damage increment (i.e., d \Delta D), which will come (out) from these methods.

**Reply by batodon** [reply]
Hi @GiudGiud, I believe I finally found a way to update the damage variable with the scalar value. I overrode the iterationFinalize() method (which is inherited from the radialreturnstressupdate class), and it compiled successfully. What do you think? 

[damageEvol.zip](https://github.com/idaholab/moose/files/9680027/damageEvol.zip)

**Reply by GiudGiud** [reply]
It's odd that you dont override `computeQpProperties`.
If `iterationFinalize` is called at the right times (eg every time an element loop enters a new element), it's possibly OK to do that

**Comment by ngrilli** [comment, marked as answer]
@batodon if you want to explicitly update a material property in time, you can "declare" the material property and "get" the old material property in a material class, then you implement an explicit Euler update in computeQpProperties.

You can see a slightly complicated example here:
https://github.com/ngrilli/c_pfor_am/blob/main/src/materials/ComputeCrystalPlasticityStressDamage.C

See the variables:
_plastic_work
_plastic_work_old

that I am using for updating the plastic work in a plasticity model using:

_plastic_work[_qp] = _plastic_work_old[_qp] + variation_of_plastic_work

You can use the variable _dt for the timestep

Nicolò

**Reply by batodon** [reply]
Thanks a lot, @ngrilli. This lead was very helpful!
